### PR TITLE
Arbie & Yin - Solved Issue #2 & #3

### LINKED ISSUE #2: Issue 2: Add the Cflat Compiler to Compiler-explorer
Final version will be that users can run Cflat compiler on compiler explorer (CE) on local machines.

### LINKED ISSUE #3: Issue 3: Add the Cflat Language to Compiler-explorer
On the left side of CE, it will display the default C-flat codes, and after selecting Cflat as language and compiler, the right side will display its assembly codes.

### PR DESCRIPTION
**Pull request by arbie333**
Successfully added Cflat language and compiler to CE locally.
(Displayed Cflat input and output on local compiler explorer web page.)

### CODE REVIEW

File: lib/base-compiler.ts [hunk 1]
```diff
@@ -2185,7 +2185,7 @@ export class BaseCompiler implements ICompiler {
 
     doTempfolderCleanup(buildResult) {
         if (buildResult.dirPath && !this.delayCleanupTemp) {
-            fs.remove(buildResult.dirPath);
+            //    fs.remove(buildResult.dirPath);  Remember to undo this change before opening a PR!
```
Comment (maemre) [inline lib/base-compiler.ts]: undo this change, as you say in the comments

File: examples/cflat/default.lir [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  t:
+    $ret x
+  f:
+    $ret y
+
+  entry:
+    a = $cmp neq x y
+    $branch a t f
+}
```
Comment (maemre) [inline examples/cflat/default.lir]: Have a newline at the end of this file

File: etc/config/cflat.defaults.properties [hunk 3]
```diff
@@ -0,0 +1,4 @@
+compilers=cfalt
+compilerType=cflat
+compiler.cflat.name=cflat
+compiler.cflat.exe=/Users/arbie/Documents/CS590/farrago-master/target/debug/lirc
```
Comment (maemre) [inline etc/config/cflat.defaults.properties]: Your particular executable path shouldn't be committed to the repo. You can have it in your local properties file.

File: static/modes/cflat-mode.ts
Comment (maemre) [inline static/modes/cflat-mode.ts:1]: This is fine for now, you'll need to change this later once the concrete syntax is fully decided because it doesn't capture Cflat syntax yet.

**Review by maemre — CHANGES_REQUESTED** [review]
It looks good overall with minor changes. See my comments about small issues to fix. Also, I didn't see where the tests you have are. I'll approve this once there is a test suite for the Cflat compiler and the minor issues are fixed.

File: etc/config/cflat.defaults.properties (hunk 3 above)
Comment (arbie333) [reply etc/config/cflat.defaults.properties]: What should the execution route of compiler.cflat.exe be for cflat.defaults.properties?

File: test/cflat-tests.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        processed.should.have.property('asm');
+        const asmSegments = (processed as {asm: ParsedAsmResultLine[]}).asm;
+        asmSegments.should.deep.equal(expectedSegments);
+    }
+
+    it('should handle errors', async () => {
+        const result = {
+            asm: '<Compilation failed>',
+        };
+        (await compiler.processAsm(result, this.filters, this.options)).should.deep.equal([
```
Comment (geniuseva) [inline test/cflat-tests.ts]: There is an error here: "filters" and "options" are undefined. How can we call them? @maemre

File: test/cflat/branch/cflatp-branch.asm [hunk 5]
```diff
@@ -0,0 +1 @@
+Compiled from "example.lir"
```
Comment (arbie333) [inline test/cflat/branch/cflatp-branch.asm]: We noticed that in this directory we needed to create a intermediate asm file between input and output file, but we do not know how to write the content. @maemre

Comment (maemre) [reply test/cflat/branch/cflatp-branch.asm]: What do you mean by an intermediate assembly file? The compiler goes from LIR code to assembly, there is no intermediate output. If you need an expected output, you can generate it by running the compiler on the input file as usual.

File: test/cflat-tests.ts (hunk 4 above)
Comment (maemre) [reply test/cflat-tests.ts]: I left some comments also on the commit. I don't understand what you're trying to test here. What's the point of calling `processAsm`? You can log the `this` object to see which properties it has.

File: etc/config/cflat.defaults.properties (hunk 3 above)
Comment (maemre) [reply etc/config/cflat.defaults.properties]: It can be just `lirc` for now, or if not having that property works, you can just remove it.

File: test/cflat-tests.ts (hunk 4 above)
Comment (geniuseva) [reply test/cflat-tests.ts]: We couldn't see your comments on the commit, where are they? 
Also, we call processAsm because we followed the way that Java test handles error. The purpose of this code is likely to test that, when given certain inputs (result, this.filters, and this.options), the processAsm function on the compiler object returns an error message in the expected format. That's how we decide to handle the case of `it('should handle errors', async () => {`.
Since eventaully the expected output is an array with a single object with the properties text equal to '<Compilation failed>' and source equal to null, can we just replace `this.filters` and `this.options` with `null`?

File: test/cflat/branch/cflatp-branch.asm (hunk 5 above)
Comment (arbie333) [reply test/cflat/branch/cflatp-branch.asm]: Since we took java-test as reference, for each test example, Java had an input file, an intermediate asm file, and a output asm file. So we thought we needed one also? Could you take a look at the Java's test folder and tell us why they need this intermediate asm file? We are confused by the usage of these files as well.
(We had already run the compiler on the input file and got the expected output file.)

Comment (maemre) [reply test/cflat/branch/cflatp-branch.asm]: In case of Java, those files are there because of the particulars of the compilation steps the java compiler implementation uses, and the `processAsm` function is specialized for Java in that case. Those files aren't intermediate files, but they seem like the output files for each class (because Java generates each `.class` file separately). You can see that from other Java tests. You don't need such mechanisms for Cflat because there is one output file. All you need is an expected output file, and calling the compiler will generate the actual output which should be (near-)identical.

File: test/cflat-tests.ts (hunk 4 above)
Comment (maemre) [reply test/cflat-tests.ts]: > We couldn't see your comments on the commit, where are they?

They also appear on the PR. You can click on the particular commit and load the relevant files to see the comments.

I don't think you even need a `processAsm` function, you don't need to test it if you aren't overloading it. You just need to test that the resulting output file is what the compiler produces.

**Comment by arbie333** [comment]
@maemre We finished everything for Cflat test based on your comments.

**Review by maemre — APPROVED** [review]

**Comment by maemre** [comment]
@geniuseva @arbie333 I approved this, you can merge it after our meeting today if the tests are passing
